crates/ruff_python_semantic/src/red_knot/types/infer.rs

```diff
@@ -0,0 +1,980 @@
+                            ast::Operator::Mod => n
+                                .checked_rem(m)
+                                .map(Type::IntLiteral)
+                                // TODO division by zero error
+                                .unwrap_or(Type::Unknown),
+                            _ => todo!("complete binop op support for IntLiteral"),
+                        }
+                    }
+                    _ => todo!("complete binop right_ty support for IntLiteral"),
+                }
+            }
+            _ => todo!("complete binop support"),
+        }
+    }
+
+    fn lower_type_parameters(&mut self, _type_parameters: &TypeParams) {
+        todo!("Lower type parameters")
+    }
+
+    pub(super) fn finish(mut self) -> TypeInference {
+        let symbol_tys: IndexVec<_, _> = self
+            .index
+            .symbol_table(self.file_scope_id)
+            .symbol_ids()
+            .map(|symbol| {
+                let ty = self.local_definition_ty(symbol);
+                ty.into_type(&mut self)
+            })
+            .collect();
+
+        self.types.symbol_tys = symbol_tys;
+        self.types.shrink_to_fit();
+        self.types
+    }
+
+    fn union_ty(&mut self, ty: UnionType) -> Type {
+        Type::Union(TypeId {
+            scope: self.scope,
+            local: self.types.union_types.push(ty),
+        })
+    }
+
+    fn function_ty(&mut self, ty: FunctionType) -> Type {
+        Type::Function(TypeId {
+            scope: self.scope,
+            local: self.types.function_types.push(ty),
+        })
+    }
+
+    fn class_ty(&mut self, ty: ClassType) -> Type {
+        Type::Class(TypeId {
+            scope: self.scope,
+            local: self.types.class_types.push(ty),
+        })
+    }
+
+    fn typing_context(&self) -> TypingContext {
+        TypingContext::local(self.db, self.scope, &self.types)
+    }
+
+    fn local_definition_ty(&mut self, symbol: ScopeSymbolId) -> DefinitionType {
+        let symbol = self.symbol_table.symbol(symbol);
+        let mut definitions = symbol
+            .definitions()
+            .iter()
+            .filter_map(|definition| self.definition_tys.get(definition).copied());
+
+        let Some(first) = definitions.next() else {
+            return DefinitionType::Type(Type::Unbound);
+        };
+
+        if let Some(second) = definitions.next() {
+            let context = self.typing_context();
+            let mut builder = UnionTypeBuilder::new(&context);
+            builder = builder.variant(first).variant(second);
+
+            for variant in definitions {
+                builder = builder.variant(variant);
+            }
+
+            // The fact that the interner is local to a body means that we can't reuse the same union type
+            // across different call sites. But that's something we aren't doing yet anyway. Our interner doesn't
+            // deduplicate union types that are identical.
```

Comment:
I'm not sure I understand this comment. Is it really just saying we'll only be able to de-duplicate per scope, rather than per-file?

De-duplication is something that I definitely think we should add, so it would be good to have a TODO comment (or an issue) for that.